### PR TITLE
[#3827] Clean temporary directory on OSX.

### PR DESCRIPTION
**Pull request by brunogola**
Scope
=====

On OSX the temporary pip directory was not removed if pip failed as it is not under /tmp.

Changes
=======

Add code to clean the folder under ${TMPDIR}.

How to try and test the changes
===============================

reviewers: @adiroiban @dumol 

Please check if changes make sense.

### CODE REVIEW

**Comment by brunogola** [comment]
needs-review

**Comment by brunogola** [comment]
needs-review

**Comment by adiroiban** [comment]
don't forget the ticket id in the title.

needs-review

File: paver.sh [hunk 1]
```diff
@@ -99,6 +99,12 @@ clean_build() {
     # In some case pip hangs with a build folder in temp and
     # will not continue until it is manually removed.
     rm -rf /tmp/pip*
+
+    # On the OSX build server tmp is in $TMPDIR
+    # check if TMPDIR is set before trying to clean it.
+    if [ ! -z "${TMPDIR-}" ]; then
+        rm -rf ${TMPDIR}/pip*
+    fi
```
Comment (adiroiban) [inline paver.sh]: maybe and an else here and put the above ` rm -rf /tmp/pip*` ... so that when we have TMPDIR we only delete TMPDIR and not other folders

File: paver.sh [hunk 2]
```diff
@@ -99,6 +99,12 @@ clean_build() {
     # In some case pip hangs with a build folder in temp and
     # will not continue until it is manually removed.
     rm -rf /tmp/pip*
+
+    # On the OSX build server tmp is in $TMPDIR
```
Comment (adiroiban) [inline paver.sh]: this comment  should be inside the block which deals with the osx condition

**Comment by adiroiban** [comment]
looks good. ... only a few minor comments

changes-approved

File: paver.sh [hunk 3]
```diff
@@ -98,7 +98,13 @@ clean_build() {
     fi
     # In some case pip hangs with a build folder in temp and
     # will not continue until it is manually removed.
-    rm -rf /tmp/pip*
+    # On the OSX build server tmp is in $TMPDIR
+    if [ ! -z "${TMPDIR-}" ]; then
```
Comment (dumol) [inline paver.sh]: Two minor questions/suggestions:
  * why use `! -z` instead of `-n`?
  * why use parameter expansion when we are not interested in corner cases here?

Comment (brunogola) [inline paver.sh]: `-n`  works. 

the parameter expansion is used because if we test only for `${TMPDIR}` we might get 'unbound variable' error when it is not defined.

thanks!

Comment (dumol) [inline paver.sh]: Not sure what you mean… I tried the `bash` in the AIX 5.3 build slave, the oldest we've got:
```
-bash-3.2$ bash --version
GNU bash, version 3.2.16(1)-release (powerpc-ibm-aix5.2.0.0)
Copyright (C) 2005 Free Software Foundation, Inc.
-bash-3.2$ if [ -n "$NOSUCHDIR" ]; then echo 'we have $NOSUCHDIR!'; fi
-bash-3.2$ export NOSUCHDIR="/tmp"
-bash-3.2$ if [ -n "$NOSUCHDIR" ]; then echo 'we have $NOSUCHDIR!'; fi
we have $NOSUCHDIR!
```
I think using parameter expansion here is overkill and will confuse us later.

Comment (brunogola) [inline paver.sh]: actually for some reason `-n` would give us a false positive:

```
[bgola@charango noisegrid]$ echo ${xx}
bash: xx: unbound variable
[bgola@charango noisegrid]$ if [ ! -z ${xx-} ]; then echo "not set"; fi
[bgola@charango noisegrid]$ if [ -n  ${xx-} ]; then echo "not set"; fi
S
```

so I will keep `! -z`

Comment (dumol) [inline paver.sh]: Because you use two tricks and they work together. But the easiest thing would be to use no trick at all…

Comment (brunogola) [inline paver.sh]: our script runs with `set -u`, so:

```
$ if [ -n "$NOSUCHDIR" ]; then echo 'we have $NOSUCHDIR!'; fi
bash: NOSUCHDIR: unbound variable
```

Comment (dumol) [inline paver.sh]: Ah, true, I forgot about the draconian settings of `paver.sh`. This is because of `set -o nounset`. Parameter extensions is indeed the solution for that, I have also used it at https://github.com/chevah/python-package/blob/master/python-modules/chevah-python-test/get_binaries_deps.sh#L19

Interestingly the `-n` and `-z` don't work equally here… I looked at the manual and couldn't figure it out. But is seems as though `+` would be better than `-` here, from what I gather at https://www.gnu.org/software/bash/manual/bashref.html#Shell-Parameter-Expansion, as `+` doesn't substitute anything if the parameter is null or unset. However, in my tests, both work for this check in the bash shell included with OS X.
